File: lib/src/db/entities/dashboard-panel.entity.ts
import { Column, Entity, ManyToOne, PrimaryGeneratedColumn } from "typeorm";
import { BaseEntity } from "../../common/base/base.entity";
import { Dashboard } from "./dashboard.entity";
import { Dashboard as IDashboard, DashboardPanel as IDashboardPanel, PANEL_TYPE, PanelConfiguration, PanelGridPosition } from "@traceo/types";

@Entity()
export class DashboardPanel extends BaseEntity implements IDashboardPanel {
    @PrimaryGeneratedColumn("uuid")
    id?: string;

    @Column({
        type: "varchar",
        nullable: false
    })
    title: string;

    @Column({ type: "varchar", nullable: true })
    description: string;

    @Column({ type: "varchar", nullable: true })
    type: PANEL_TYPE;

    @Column({
        type: "simple-json",
        nullable: false
    })
    gridPosition: PanelGridPosition;

    @Column({ type: "simple-json", nullable: false })
    config: PanelConfiguration;

    @ManyToOne(() => Dashboard, dashboard => dashboard.panels, {
        onUpdate: "CASCADE",
        onDelete: "CASCADE"
    })
    dashboard: IDashboard;
}

File: server/traceo-common/traceo-common-jpa/src/main/java/org/traceo/common/jpa/entities/DashboardPanelEntity.java
package org.traceo.common.jpa.entities;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;
import org.traceo.common.transport.dto.PanelConfigurationDto;
import org.traceo.common.transport.dto.PanelGridPosition;
import org.traceo.common.transport.dto.api.DashboardDto;
import org.traceo.common.transport.dto.api.DashboardPanelDto;

import java.util.Objects;

@Entity
@Table(name = "traceo_dashboard_panel")
@Getter
@Setter
public class DashboardPanelEntity extends BaseEntity {
    @Column(nullable = false)
    private String title;

    private String description;

    @Column(nullable = false, name = "grid_position")
    @JdbcTypeCode(value = SqlTypes.JSON)
    private PanelGridPosition gridPosition;

    @Column(nullable = false)
    @JdbcTypeCode(value = SqlTypes.JSON)
    private PanelConfigurationDto config;

    @ManyToOne(cascade = {
            CascadeType.REMOVE,
            CascadeType.PERSIST
    })
    private DashboardEntity dashboard;

    public static DashboardPanelEntity mapToEntity(DashboardPanelDto dto, DashboardEntity dashboard) {
        DashboardPanelEntity entity = new DashboardPanelEntity();
        entity.setDashboard(dashboard);
        entity.setGridPosition(dto.getGridPosition());
        entity.setTitle(dto.getTitle());
        entity.setDescription(dto.getDescription());
        entity.setConfig(dto.getConfig());
        return entity;
    }

    public static DashboardPanelDto mapToModel(DashboardPanelEntity entity) {
        DashboardPanelDto dto = new DashboardPanelDto();
        dto.setGridPosition(entity.getGridPosition());
        dto.setTitle(entity.getTitle());
        dto.setDescription(entity.getDescription());
        dto.setConfig(entity.getConfig());

        if (!Objects.isNull(entity.getDashboard())) {
            dto.setDashboardId(entity.getDashboard().getId());
        }

        return dto;
    }
}
